experimental/oak_baremetal_channel/SPEC.MD

```diff
@@ -0,0 +1,118 @@
+# Oak Baremetal Communication Protocol Specification
+
+This document serves as a specification of the protocol used for communication
+between the Untrusted Loader and Trusted Runtime. The Trusted Runtime should be
+able to communicate with any implementation of the Untrusted Loader that
+implements this specification.
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD",
+"SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be
+interpreted as described in [RFC 2119](https://www.rfc-editor.org/rfc/rfc2119).
+
+## Method
+
+Methods are exposed by the Trusted Runtime. They MAY be remotely invoked by the
+Unrusted Loader. Each method has clearly defined parameters and return types.
+The Oak IDL is used to define the methods exposed by the Trusted Runtime.
+
+## Invocation
+
+An invocations of a Method. Each invocation consists out of a Request Message
+followed by a Response Message.
+
+### Issuing an invocation
+
+The Untrusted Loader MAY issue an invocation by sending a Request Message.
+
+The Trusted Runtime MUST NOT issue invocations. It MUST ignore any Response
+Messages it receives.
+
+### Handling an invocation
+
+The Trusted Runtime MUST respond to each Request Message with a single Response
+Message. It SHOULD respond to Request Messages in first-in-first-out order, but
+MAY prioritize some requests. The Trusted Runtime MAY issue Response Messages
+before the Request Message has been received in its entirety. Sending a Response
+Messages concludes the invocation.
+
+The Untrusted Loader MUST NOT send Response Messages. It MUST ignore any Request
+Messages it receives.
+
+## Messages
+
+A message forms half of an Invocation.
+
+### Request Message Definition
+
+A byte encoded request message MUST consist of the following seqeuential fields:
+
+- length, u32, little endian: length of the entire message, including this field
+- invocation_id, u32, little endian: Identifies the invocation the request
+  initiates
+- method_id, u32, little endian: identifies the method that is being invoked
+- padding, 4 bytes: reserved to maintain 8 byte alignment. MUST be unset and
+  ignored
+- body, variable length byte array: the method’s arguments, encoded with
+  flatbuffer
+
+### Response Message Definition
+
+A byte encoded response message MUST consist of the following seqeuential
+fields:
+
+- length, u32, little endian: length of the entire message, including this field
+- invocation_id, u32, little endian: MUST match the invocation_id of the
+  relevant Request Message
+- status_code, u32, little endian: response status as a
+  [gRPC status code](https://grpc.github.io/grpc/core/md_doc_statuscodes.html)
+- status_message_length, u32, little endian: length of the status message
+- status_message, utf-8 encoded string: developer-facing error message
+- body, variable length byte array: the method’s return value, encoded with
+  flatbuffer
```

Comment:
Good question! I was thinking about this too, and I don't lean too strongly either way. 

The reason I went with the current approach is that this is [what gRPC does](https://github.com/grpc/grpc/blob/master/src/proto/grpc/status/status.proto), and we use gRPC status codes.

Reply:
Doesn't gRPC convert this status message into the response body though?

Reply:
Looked into a bit more, and while gRPC does not use the body to store the status message, it does seem to omit it for non error responses. :) https://github.com/grpc/grpc/blob/master/doc/PROTOCOL-HTTP2.md#errors

Which means we probably could indeed follow your suggestion. The only downside would be that it adds some coupling between the gRPC status messages and our message format (since we depend on gRPCs defintion of "Ok").

Reply:
Updated to store it in the body, but happy to revert to using a dedicated field if you think it'd be better. :P

Reply:
Either way, I would avoid having a len-delimited field inside an already len-delimited message.

Personally I think we can just say if the status is OK (whatever status code we decide to represent that), then the body contains the response; otherwise, the body contains the error message supporting the non-OK status.

Reply:
implemented like that :)